bflb: attempt to address e24 cores boot failures

VynDragon opened the pull request:
Clean up code and add attempts to prevent bl60x and bl70x from failing to boot

nandojve commented on soc/bflb/bl60x/Kconfig.defconfig:
```diff
@@ -5,9 +5,9 @@
+# 1000, 100, 10, 1
 config SYS_CLOCK_TICKS_PER_SEC
-	default 5000
+	default 1000
```
Ideally this could came from devicetree. Then user can select the Tick there.

VynDragon replied:
How would one do that? I havent seen it done from DT yet, another property like the timebase one?

nandojve replied:
We use the `riscv,machine-timer` to tick the system but that bind does not have the TICKS_PER_SEC int property. We could add this there. Then you can use the devicetree Kconfig extensions to pick the value and set the CONFIG_SYS_CLOCK_TICKS_PER_SEC.

I would add a help entry to tell what are the valid numbers and drop the comments.

VynDragon replied:
That's not related to the machine timer here, this is a zephyr setting (that applies to all platforms) for the kernel ticks.

nandojve replied:
So, why do we use machine timer ?

VynDragon replied:
For the counting of the time on riscv platforms? eg this is a kernel setting vs the other being a peripheral/hardware property/setting

josuah replied:
It seems like other SoCs could also benefit from it.
Would it be worth splitting this suggested change into a different PR with goal to reduce fragmentation?

https://github.com/search?q=repo%3Azephyrproject-rtos%2Fzephyr+config+SYS_CLOCK_TICKS_PER_SEC&type=code

I remember I tried to use the devicetree for SYS_CLOCK_TICKS_PER_SEC on another RISC-V platform but did not get it work. Maybe time for me to try that again.

VynDragon replied:
It should definitely not be in this PR.

nandojve replied:
> It should definitely not be in this PR.

I'm not asking to be made in this PR but I'm raising the topic because I think it seems to be relevant.

josuah commented on soc/bflb/bl70x/Kconfig.defconfig:
```diff
@@ -5,9 +5,9 @@
 # On BL702, BL704 and BL706, not all values work, here is a list of some that work:
-# 2000, 1000, 100, 10, 1
+# 1000, 100, 10, 1
 config SYS_CLOCK_TICKS_PER_SEC
-	default 2000
+	default 1000
```
In both case, it seems like setting the `SYS_CLOCK_TICS_PER_SEC` above 1000 ticks per second is causing troubles...

```
  .  time unit
  |  sys tick irq

2000 ticks/sec ....|....|....|....|....|....|....
1000 ticks/sec .........|.........|.........|....
                   ^
                  one possible difference is the absence of interrupt
                  early-on in the boot process
```

Is the systick modification also helping with getting things to work?

VynDragon replied:
That is possible.

josuah commented on soc/bflb/common/e24/cache.c:
```diff
@@ -0,0 +1,50 @@
+{
+	uint32_t tmp;
+
+	tmp = *((uint32_t *)(L1C_BASE + L1C_CONFIG_OFFSET));
```
I once got the compiler optimizing away a raw memory write when the `volatile` keyword was not added:

`tmp = *((volatile uint32_t *)(L1C_BASE + L1C_CONFIG_OFFSET));`

Or using `<zephyr/sys/sys_io.h>` which has the `volatile`:

`tmp = sys_read32(L1C_BASE + L1C_CONFIG_OFFSET);`

It does not seem to have any downside anywhere here though...

VynDragon replied:
sys_read can not be used here as it is in flash. Otherwise I havent had that issue yet but i guess this is a good safety?

VynDragon replied:
done

josuah reviewed (commented):
At bootup, plenty of RAM types are containing random data.
Maybe this is the same for the caches, and they need to be flushed at startup?

This might be already done as part of chip init, except this could be taking time and starting the hardware right away could end-up executing random bytes?

If so, an interrupt a bit earlier (like when systick is faster than 1000/sec) could end-up executing some region of flash that has cache not flushed yet...

Not sure what that could be really!

VynDragon commented on soc/bflb/common/e24/cache.c:
```diff
@@ -0,0 +1,50 @@
+
+	tmp = *((uint32_t *)(L1C_BASE + L1C_CONFIG_OFFSET));
+	tmp |= 1 << L1C_BYPASS_POS;
+	tmp &= ~(1 << L1C_CACHEABLE_POS);
+	tmp &= ~L1C_WAY_DIS_MSK;
+	tmp &= ~(1 << L1C_CNT_EN_POS);
+	*((uint32_t *)(L1C_BASE + L1C_CONFIG_OFFSET)) = tmp;
+
+	tmp &= ~(1 << L1C_INVALID_EN_POS);
+	*((uint32_t *)(L1C_BASE + L1C_CONFIG_OFFSET)) = tmp;
+	__asm__ volatile (".rept 4 ; nop ; .endr");
+
+	tmp |= (1 << L1C_INVALID_EN_POS);
+	*((uint32_t *)(L1C_BASE + L1C_CONFIG_OFFSET)) = tmp;
+	__asm__ volatile (".rept 4 ; nop ; .endr");
+
+	while ((*((uint32_t *)(L1C_BASE + L1C_CONFIG_OFFSET)) & L1C_INVALID_DONE_MSK) == 0) {
+		__asm__ volatile (".rept 50 ; nop ; .endr");
+	}
+
+	tmp = *((uint32_t *)(L1C_BASE + L1C_CONFIG_OFFSET));
+	tmp |= 1 << L1C_BYPASS_POS;
+	*((uint32_t *)(L1C_BASE + L1C_CONFIG_OFFSET)) = tmp;
+
+	tmp &= ~(1 << L1C_BYPASS_POS);
+	tmp |= 1 << L1C_CNT_EN_POS;
+	*((uint32_t *)(L1C_BASE + L1C_CONFIG_OFFSET)) = tmp;
+
+	tmp &= ~L1C_WAY_DIS_MSK;
+	*((uint32_t *)(L1C_BASE + L1C_CONFIG_OFFSET)) = tmp;
```
L1C_WAY_DIS_MSK is a disabling mask (1 = disabled, 0 = enabled), both writes to that are ensuring they are all enabled, the first one probably is not required, it defaults to the right number of disablement (none, datasheet is wrong of bootrom sets them all to 0) anyway.

VynDragon replied:
Actually it's probably to ensure all 4 cache ways are invalidated, dont remember, this bit of code was tested and written over a year ago.

VynDragon commented on soc/bflb/common/e24/intc_clic.S:
removed some more stuffs.

josuah reviewed (dismissed):
Thank you for the explanations.

nandojve commented on soc/bflb/bl60x/Kconfig:
```diff
@@ -15,7 +15,6 @@ config SOC_SERIES_BL60X
 	select RISCV
 	select RISCV_HAS_CLIC
 	select RISCV_MACHINE_TIMER
-	select RISCV_SOC_HAS_CUSTOM_IRQ_HANDLING
```
This should be in a separate commit. We are dropping this in favor of what ?

VynDragon replied:
separated the changes related to the interrupt cleanup in their own commit.

nandojve commented on soc/bflb/bl60x/soc.c:
```diff
@@ -21,6 +21,8 @@
 #include <hbn_reg.h>
 #include <pds_reg.h>
 
+extern void soc_bflb_cache_invalidate(void);
```
Cache should be in another commit with explanation about why it is necessary ? Why RISC-V arch infrastructure do not handle this ?

VynDragon replied:
That *is* the thing the commit description is about... Because there is as many riscv cache schemes that there are riscv SoCs.

VynDragon replied:
okay no seems it need to go in drivers/cache and enable CONFIG_CACHE_MANAGEMENT...

VynDragon replied:
Can we have this for now and leave making a driver for all the caches for later?

VynDragon replied:
Or maybe i can introduce it with the DMA.

nandojve commented on soc/bflb/bl70x/Kconfig.defconfig:
```diff
@@ -5,9 +5,9 @@
 
 # On BL702, BL704 and BL706, not all values work, here is a list of some that work:
-# 2000, 1000, 100, 10, 1
+# 1000, 100, 10, 1
```
Why are we changing this ? Another commit with clear explanation.

VynDragon replied:
> it seems that you fix the problem but it was a guess because the technical details are not present

I thought I made it obvious enough this was the case here. It fixes the issue, which is critical, and I dont know why.

nandojve commented on soc/bflb/common/e24/intc_clic.S:
```diff
@@ -9,19 +9,6 @@
  */
 
 #include <zephyr/arch/cpu.h>
-#include "clic.h"
```
I'm wondering why we still need this file after all these removals.

VynDragon replied:
You are right, it's not needed. But it triggers the bug this PR is trying to fix when using the riscv-privileged .S... Which I didnt catch at first and thought it was because it needed it defined.

VynDragon committed:
soc: bflb: Remove Extraneous code

Remove unecessary unused code and Kconfig

Signed-off-by: Camille BAUD <mail@massdriver.space>

VynDragon committed:
soc: bflb: address e24 boot failures

This should fix e24 failing to boot by ensuring
interrupts are disabled at startup

Signed-off-by: Camille BAUD <mail@massdriver.space>

josuah reviewed (approved):
Waiting a bit before enabling IRQs seems present on other RISC-V:

https://github.com/zephyrproject-rtos/zephyr/blob/209bfbde89710edd83f075c258c492f646f64565/drivers/interrupt_controller/intc_vexriscv_litex.c#L97-L98

This is what the SDK seems to be doing too, in [`SystemInit()`](https://github.com/bouffalolab/bouffalo_sdk/blob/811e8c1f1cd62695cedd47367f0bf2c624707bc6/drivers/soc/bl602/std/startup/system_bl602.c#L33), disabling IRQs first:

```c
    /* global IRQ disable */
    __disable_irq();
```

IIRC there is a ROM bootloader starting first? Maybe this is what enables the IRQs to begin with.
